### ISSUE TITLE
must also pass R CMD BiocCheck with no errors and no warnings

### ISSUE THREAD

**Issue by raysinensis** [rnabioco/clustifyr#313]

**Comment by raysinensis** [comment]
$error
[1] "Invalid package Version, see http://www.bioconductor.org/developers/how-to/version-numbering/"                  
[2] "No biocViews terms found."                                                                                      
[3] "At least 80% of man pages documenting exported objects must have runnable examples. The following pages do not:"
[4] "Maintainer must register at the support site; visit https://support.bioconductor.org/accounts/signup/ ."        

$warning
[1] "Update R version dependency from 3.2.0 to 3.6."                                                                  
[2] "The following files are over 5MB in size: '.git/objects/pack/pack-d9cd4d0ae05a0ee38f136a78ebf857869170c765.pack'"
[3] "Import stats in DESCRIPTION as well as NAMESPACE."                                                               
[4] " Use TRUE/FALSE instead of T/F"                                                                                  
[5] " Avoid class() == or class() != ; use is() or !is()"                                                             
[6] " Remove set.seed usage in R code"

**Comment by kriemo** [comment]
``` r
BiocCheck::BiocCheck("~/src/rnabioco/clustifyr/")
#> This is BiocCheck version 1.20.0. BiocCheck is a work in progress.
#> Output and severity of issues may change. Installing package...
#> * Checking Package Dependencies...
#> * Checking if other packages can import this one...
#> * Checking to see if we understand object initialization...
#> * Checking for deprecated package usage...
#> * Checking for remote package usage...
#> * Checking version number...
#> * Checking version number validity...
#>     * ERROR: Invalid package Version, see
#>       http://www.bioconductor.org/developers/how-to/version-numbering/
#> * Checking R Version dependency...
#>     * WARNING: Update R version dependency from 3.2.0 to 3.6.
#> * Checking package size...
#>         Skipped... only checked on source tarball
#> * Checking individual file sizes...
#>     * WARNING: The following files are over 5MB in size:
#>       '.git/objects/pack/pack-078f3c3b7f15a4045c5fe2e8260facad63f8d458.pack
#>       .git/objects/pack/pack-eab16365fb28e2b3497f0537064324a8a70c3d60.pack'
#> * Checking biocViews...
#> * Checking that biocViews are present...
#>     * ERROR: No biocViews terms found.
#> See http://bioconductor.org/developers/how-to/biocViews/
#> * Checking build system compatibility...
#> * Checking for blank lines in DESCRIPTION...
#> * Checking if DESCRIPTION is well formatted...
#> * Checking for whitespace in DESCRIPTION field names...
#> * Checking that Package field matches directory/tarball name...
#> * Checking for Version field...
#> * Checking for valid maintainer...
#> * Checking DESCRIPTION/NAMESPACE consistency...
#>     * WARNING: Import stats in DESCRIPTION as well as NAMESPACE.
#> * Checking vignette directory...
#>     This is an unknown type of package
#> * Checking library calls...
#> * Checking for library/require of clustifyr...
#> * Checking coding practice...
#>     * NOTE: Avoid sapply(); use vapply()
#>       Found in files:
#>         compare_genelist.R (line 198, column 13)
#>         compare_genelist.R (line 234, column 17)
#>         compute_similarity.R (line 191, column 3)
#>         run_fgsea.R (line 60, column 23)
#>         run_fgsea.R (line 68, column 23)
#>         utils.R (line 132, column 12)
#>         utils.R (line 258, column 17)
#>         utils.R (line 605, column 5)
#>         utils.R (line 612, column 5)
#>         utils.R (line 619, column 5)
#>         utils.R (line 669, column 10)
#>         utils.R (line 1405, column 12)
#>         utils.R (line 1521, column 13)
#>         utils.R (line 1547, column 15)
#>         utils.R (line 1548, column 25)
#>         utils.R (line 1575, column 17)
#>         utils.R (line 1708, column 9)
#>     * NOTE: Avoid 1:...; use seq_len() or seq_along()
#>       Found in files:
#>         compare_genelist.R (line 92, column 44)
#>         compare_genelist.R (line 100, column 51)
#>         compare_genelist.R (line 253, column 25)
#>         compute_similarity.R (line 138, column 13)
#>         utils.R (line 227, column 15)
#>         utils.R (line 414, column 40)
#>         utils.R (line 416, column 61)
#>         utils.R (line 510, column 13)
#>         utils.R (line 567, column 50)
#>         utils.R (line 1301, column 24)
#>         utils.R (line 1311, column 30)
#>         utils.R (line 1313, column 27)
#>         utils.R (line 1361, column 13)
#> Warning in readLines(infile): incomplete final line found on '/Users/
#> kriemo/src/rnabioco/clustifyr/tests/testthat/test_cor.R'
#> Warning in readLines(infile): incomplete final line found on '/Users/
#> kriemo/src/rnabioco/clustifyr/tests/testthat/test_list.R'
#> Warning in readLines(infile): incomplete final line found on '/Users/
#> kriemo/src/rnabioco/clustifyr/tests/testthat/test_plots.R'
#> Warning in readLines(infile): incomplete final line found on '/Users/
#> kriemo/src/rnabioco/clustifyr/tests/testthat/test_utils.R'
#>     * WARNING: Use TRUE/FALSE instead of T/F
#>       Found in R/ directory functions:
#>         find_rank_bias()
#>         pos_neg_marker()
#>       Found in files:
#>         data-raw/cbmc_m.R
#>         tests/testthat/test_list.R
#>         tests/testthat/test_plots.R
#>         tests/testthat/test_utils.R
#>         manuscript/figs.Rmd
#>         README.Rmd
#>         vignettes/benchmark.Rmd
#>         vignettes/clustifyR.Rmd
#>         vignettes/references.Rmd
#>     * WARNING: Avoid class() == or class() != ; use is() or !is()
#>       Found in files:
#>         R/utils.R (line 82)
#>         R/utils.R (line 665)
#>         R/utils.R (line 835)
#>         R/utils.R (line 932)
#>         R/utils.R (line 1020)
#>         R/utils.R (line 1513)
#>         R/utils.R (line 1574)
#>         R/utils.R (line 1699)
#>         R/utils.R (line 1701)
#>         R/utils.R (line 1703)
#>         R/main.R (line 173)
#>         R/seurat_wrapper.R (line 31)
#>         R/seurat_wrapper.R (line 64)
#>         R/seurat_wrapper.R (line 67)
#>         R/seurat_wrapper.R (line 117)
#>         R/seurat_wrapper.R (line 142)
#>         R/seurat_wrapper.R (line 191)
#>         tests/testthat/test_utils.R (line 813)
#>         tests/testthat/test_utils.R (line 1027)
#>         tests/testthat/test_plots.R (line 180)
#>         tests/testthat/test_gsea.R (line 90)
#>     * WARNING: Remove set.seed usage in R code
#>       Found in R/ directory functions:
#>         downsample_matrix()
#>         overcluster()
#>         overcluster_test()
#> * Checking parsed R code in R directory, examples, vignettes...
#>     Found <<- in R/main.R (line 773, column 15)
#>     * NOTE: Avoid '<<-' if possible (found in 1 files)
#> * Checking function lengths....................................................................
#>     * NOTE: Recommended function length <= 50 lines.
#>       There are 21 functions > 50 lines.
#>       The longest 5 functions are:
#>         clustify.default() (R/main.R, line 79): 155 lines
#>         clustify_lists.default() (R/main.R, line 511): 134 lines
#>         clustify_nudge.default() (R/utils.R, line 765): 124 lines
#>         plot_tsne() (R/plot.R, line 20): 116 lines
#>         compare_lists() (R/compare_genelist.R, line 167): 110
#>       lines
#> * Checking man page documentation...
#>     * ERROR: At least 80% of man pages documenting exported
#>       objects must have runnable examples. The following pages do
#>       not:
#>       calc_similarity.Rd, call_to_metadata.Rd, clustify_lists.Rd,
#>   clustifyr_methods.Rd, collapse_to_cluster.Rd, cosine.Rd,
#>   find_rank_bias.Rd, get_unique_column.Rd, get_vargenes.Rd,
#>   insert_meta_object.Rd, kl_divergence.Rd, not_pretty_palette.Rd,
#>   object_ref.Rd, plot_tsne.Rd, pretty_palette_ramp_d.Rd,
#>   pretty_palette.Rd, pretty_palette2.Rd, seurat_meta.Rd,
#>   vector_similarity.Rd
#>     * NOTE: Usage of dontrun{} / donttest{} found in man page
#>       examples.
#>       4% of man pages use one of these cases.
#>       Found in the following files:
#>         call_to_metadata.Rd
#>         insert_meta_object.Rd
#>         seurat_meta.Rd
#> * Checking package NEWS...
#> * Checking unit tests...
#> * Checking skip_on_bioc() in tests...
#> * Checking formatting of DESCRIPTION, NAMESPACE, man pages, R
#>   source, and vignette source...
#>     * NOTE: Consider shorter lines; 470 lines (5%) are > 80
#>       characters long.
#>     First 6 lines:
#>       R/common_dplyr.R:6 #' @param collapse_to_cluster if a column name ...
#>       R/common_dplyr.R:33   df_temp <- tidyr::gather(df_temp, key = !!dp...
#>       R/common_dplyr.R:34   df_temp[["type"]][df_temp$r < threshold] <- ...
#>       R/common_dplyr.R:40     df_temp[lapply(df_temp[, 1], FUN = functio...
#>       R/common_dplyr.R:42     df_temp_full <- dplyr::distinct_at(df_temp...
#>       R/common_dplyr.R:51     df_temp_full <- collapse_to_cluster(df_tem...
#>     * NOTE: Consider multiples of 4 spaces for line indents, 2544
#>       lines(28%) are not.
#>     First 6 lines:
#>       R/common_dplyr.R:26   correlation_matrix <- cor_mat
#>       R/common_dplyr.R:27   if (threshold == "auto") {
#>       R/common_dplyr.R:30   }
#>       R/common_dplyr.R:31   correlation_matrix[is.na(correlation_matrix)...
#>       R/common_dplyr.R:32   df_temp <- tibble::as_tibble(correlation_mat...
#>       R/common_dplyr.R:33   df_temp <- tidyr::gather(df_temp, key = !!dp...
#>     See http://bioconductor.org/developers/how-to/coding-style/
#>     See FormatR package:
#>       https://cran.r-project.org/web/packages/formatR/index.html
#> * Checking if package already exists in CRAN...
#> * Checking for bioc-devel mailing list subscription...
#>     * NOTE: Cannot determine whether maintainer is subscribed to
#>       the bioc-devel mailing list (requires admin credentials).
#>       Subscribe here:
#>       https://stat.ethz.ch/mailman/listinfo/bioc-devel
#> * Checking for support site registration...
#>     * ERROR: Maintainer must register at the support site; visit
#>       https://support.bioconductor.org/accounts/signup/ .
#> 
#> 
#> Summary:
#> ERROR count: 4
#> WARNING count: 6
#> NOTE count: 8
#> For detailed information about these checks, see the BiocCheck
#> vignette, available at
#> https://bioconductor.org/packages/3.9/bioc/vignettes/BiocCheck/inst/doc/BiocCheck.html#interpreting-bioccheck-output
#> BiocCheck FAILED.
#> $error
#> [1] "Invalid package Version, see http://www.bioconductor.org/developers/how-to/version-numbering/"                  
#> [2] "No biocViews terms found."                                                                                      
#> [3] "At least 80% of man pages documenting exported objects must have runnable examples. The following pages do not:"
#> [4] "Maintainer must register at the support site; visit https://support.bioconductor.org/accounts/signup/ ."        
#> 
#> $warning
#> [1] "Update R version dependency from 3.2.0 to 3.6."                                                                                                                                       
#> [2] "The following files are over 5MB in size: '.git/objects/pack/pack-078f3c3b7f15a4045c5fe2e8260facad63f8d458.pack .git/objects/pack/pack-eab16365fb28e2b3497f0537064324a8a70c3d60.pack'"
#> [3] "Import stats in DESCRIPTION as well as NAMESPACE."                                                                                                                                    
#> [4] " Use TRUE/FALSE instead of T/F"                                                                                                                                                       
#> [5] " Avoid class() == or class() != ; use is() or !is()"                                                                                                                                  
#> [6] " Remove set.seed usage in R code"                                                                                                                                                     
#> 
#> $note
#> [1] " Avoid sapply(); use vapply()"                                                                                                                                                     
#> [2] " Avoid 1:...; use seq_len() or seq_along()"                                                                                                                                        
#> [3] "Avoid '<<-' if possible (found in 1 files)"                                                                                                                                        
#> [4] "Recommended function length <= 50 lines."                                                                                                                                          
#> [5] "Usage of dontrun{} / donttest{} found in man page examples."                                                                                                                       
#> [6] "Consider shorter lines; 470 lines (5%) are > 80 characters long."                                                                                                                  
#> [7] "Consider multiples of 4 spaces for line indents, 2544 lines(28%) are not."                                                                                                         
#> [8] "Cannot determine whether maintainer is subscribed to the\nbioc-devel mailing list (requires admin credentials).  Subscribe\nhere: https://stat.ethz.ch/mailman/listinfo/bioc-devel"
```

<sup>Created on 2019-10-30 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by kriemo** [comment]
Some progress:

``` r
BiocCheck::BiocCheck("~/src/rnabioco/clustifyr/")
#> This is BiocCheck version 1.20.0. BiocCheck is a work in progress.
#> Output and severity of issues may change. Installing package...
#> * Checking Package Dependencies...
#> * Checking if other packages can import this one...
#> * Checking to see if we understand object initialization...
#> * Checking for deprecated package usage...
#> * Checking for remote package usage...
#> * Checking version number...
#> * Checking version number validity...
#>     * ERROR: Invalid package Version, see
#>       http://www.bioconductor.org/developers/how-to/version-numbering/
#> * Checking R Version dependency...
#> * Checking package size...
#>         Skipped... only checked on source tarball
#> * Checking individual file sizes...
#>     * WARNING: The following files are over 5MB in size:
#>       '.git/objects/pack/pack-078f3c3b7f15a4045c5fe2e8260facad63f8d458.pack
#>       .git/objects/pack/pack-eab16365fb28e2b3497f0537064324a8a70c3d60.pack'
#> * Checking biocViews...
#> * Checking that biocViews are present...
#> * Checking package type based on biocViews...
#>     Software
#> * Checking for non-trivial biocViews...
#> * Checking that biocViews come from the same category...
#> * Checking biocViews validity...
#> * Checking for recommended biocViews...
#> * Checking build system compatibility...
#> * Checking for blank lines in DESCRIPTION...
#> * Checking if DESCRIPTION is well formatted...
#> * Checking for whitespace in DESCRIPTION field names...
#> * Checking that Package field matches directory/tarball name...
#> * Checking for Version field...
#> * Checking for valid maintainer...
#> * Checking DESCRIPTION/NAMESPACE consistency...
#> * Checking vignette directory...
#>     This is a software package
#> * Checking library calls...
#> * Checking for library/require of clustifyr...
#> * Checking coding practice...
#>     * NOTE: Avoid sapply(); use vapply()
#>       Found in files:
#>         compare_genelist.R (line 224, column 13)
#>         compare_genelist.R (line 261, column 17)
#>         compute_similarity.R (line 202, column 3)
#>         run_fgsea.R (line 63, column 23)
#>         run_fgsea.R (line 73, column 25)
#>         utils.R (line 137, column 12)
#>         utils.R (line 268, column 17)
#>         utils.R (line 669, column 5)
#>         utils.R (line 676, column 5)
#>         utils.R (line 683, column 5)
#>         utils.R (line 736, column 10)
#>         utils.R (line 1499, column 12)
#>         utils.R (line 1620, column 13)
#>         utils.R (line 1656, column 5)
#>         utils.R (line 1663, column 5)
#>         utils.R (line 1700, column 19)
#>         utils.R (line 1847, column 9)
#>     * NOTE: Avoid 1:...; use seq_len() or seq_along()
#>       Found in files:
#>         compare_genelist.R (line 102, column 44)
#>         compare_genelist.R (line 120, column 51)
#>         compare_genelist.R (line 284, column 25)
#>         compute_similarity.R (line 149, column 13)
#>         utils.R (line 237, column 15)
#>         utils.R (line 435, column 40)
#>         utils.R (line 437, column 61)
#>         utils.R (line 533, column 13)
#>         utils.R (line 614, column 50)
#>         utils.R (line 1396, column 24)
#>         utils.R (line 1406, column 30)
#>         utils.R (line 1408, column 27)
#>         utils.R (line 1456, column 13)
#> * Checking parsed R code in R directory, examples, vignettes...
#> * Checking function lengths................................................................................
#>     * NOTE: Recommended function length <= 50 lines.
#>       There are 26 functions > 50 lines.
#>       The longest 5 functions are:
#>         clustify.default() (R/main.R, line 79): 165 lines
#>         plot_tsne() (R/plot.R, line 24): 138 lines
#>         clustify_lists.default() (R/main.R, line 539): 136 lines
#>         clustify_nudge.default() (R/utils.R, line 832): 129 lines
#>         compare_lists() (R/compare_genelist.R, line 193): 115
#>       lines
#> * Checking man page documentation...
#>     * NOTE: Consider adding runnable examples to the following man
#>       pages which document exported objects:
#>       call_to_metadata.Rd, clustifyr_methods.Rd, cosine.Rd,
#>   insert_meta_object.Rd, kl_divergence.Rd, seurat_meta.Rd,
#>   vector_similarity.Rd
#>     * NOTE: Usage of dontrun{} / donttest{} found in man page
#>       examples.
#>       4% of man pages use one of these cases.
#>       Found in the following files:
#>         call_to_metadata.Rd
#>         insert_meta_object.Rd
#>         seurat_meta.Rd
#> * Checking package NEWS...
#> * Checking unit tests...
#> * Checking skip_on_bioc() in tests...
#> * Checking formatting of DESCRIPTION, NAMESPACE, man pages, R
#>   source, and vignette source...
#>     * NOTE: Consider shorter lines; 408 lines (4%) are > 80
#>       characters long.
#>     First 6 lines:
#>       R/common_dplyr.R:154     if (!(all(unique(df_temp[[cluster_col]]) ...
#>       R/common_dplyr.R:156         "cluster_col from clustify step and j...
#>       R/common_dplyr.R:258     dplyr::filter(df_temp_full2, !!dplyr::sym...
#>       R/common_dplyr.R:270 #' @param collapse_to_cluster if a column nam...
#>       R/common_dplyr.R:306     dplyr::mutate(dplyr::group_by_at(df_temp,...
#>       R/compare_genelist.R:38 #' @param marker_df dataframe of candidate...
#>     * NOTE: Consider multiples of 4 spaces for line indents, 2763
#>       lines(29%) are not.
#>     First 6 lines:
#>       R/common_dplyr.R:27   correlation_matrix <- cor_mat
#>       R/common_dplyr.R:28   if (threshold == "auto") {
#>       R/common_dplyr.R:31   }
#>       R/common_dplyr.R:32   correlation_matrix[is.na(correlation_matrix)...
#>       R/common_dplyr.R:33   df_temp <-
#>       R/common_dplyr.R:35   df_temp <- tidyr::gather(
#>     See http://bioconductor.org/developers/how-to/coding-style/
#>     See FormatR package:
#>       https://cran.r-project.org/web/packages/formatR/index.html
#> * Checking if package already exists in CRAN...
#> * Checking for bioc-devel mailing list subscription...
#>     * NOTE: Cannot determine whether maintainer is subscribed to
#>       the bioc-devel mailing list (requires admin credentials).
#>       Subscribe here:
#>       https://stat.ethz.ch/mailman/listinfo/bioc-devel
#> * Checking for support site registration...
#>     Maintainer is registered at support site.
#> 
#> 
#> Summary:
#> ERROR count: 1
#> WARNING count: 1
#> NOTE count: 8
#> For detailed information about these checks, see the BiocCheck
#> vignette, available at
#> https://bioconductor.org/packages/3.9/bioc/vignettes/BiocCheck/inst/doc/BiocCheck.html#interpreting-bioccheck-output
#> BiocCheck FAILED.
#> $error
#> [1] "Invalid package Version, see http://www.bioconductor.org/developers/how-to/version-numbering/"
#> 
#> $warning
#> [1] "The following files are over 5MB in size: '.git/objects/pack/pack-078f3c3b7f15a4045c5fe2e8260facad63f8d458.pack .git/objects/pack/pack-eab16365fb28e2b3497f0537064324a8a70c3d60.pack'"
#> 
#> $note
#> [1] " Avoid sapply(); use vapply()"                                                                                                                                                     
#> [2] " Avoid 1:...; use seq_len() or seq_along()"                                                                                                                                        
#> [3] "Recommended function length <= 50 lines."                                                                                                                                          
#> [4] "Consider adding runnable examples to the following man pages which document exported objects:"                                                                                     
#> [5] "Usage of dontrun{} / donttest{} found in man page examples."                                                                                                                       
#> [6] "Consider shorter lines; 408 lines (4%) are > 80 characters long."                                                                                                                  
#> [7] "Consider multiples of 4 spaces for line indents, 2763 lines(29%) are not."                                                                                                         
#> [8] "Cannot determine whether maintainer is subscribed to the\nbioc-devel mailing list (requires admin credentials).  Subscribe\nhere: https://stat.ethz.ch/mailman/listinfo/bioc-devel"
```

<sup>Created on 2019-10-30 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by kriemo** [comment]
Once ready to submit we can bump the package version. File size warning looks like it is picking up files from `.git`, which shouldn't be part of the rbuild or built tarball. I think we can ignore this warning. The notes would be nice to fix if time permits.

**Comment by kriemo** [comment]
@raysinensis 

BiocCheck is failing due to vignettes with external dependencies such as `clustifyrdata`. 
https://bioconductor.org/spb_reports/clustifyr_buildreport_20191125180633.html

For Bioc submisison we should probably simplify the supplemental vignettes or move them. Perhaps we can move `benchmark.Rmd` and `references.Rmd` to `clustifyrdata`. `visualizations.Rmd` might be simplified to  just use data internal to `clustifyr` or moved to `clustifyrdata`. What do you think?

`clustifyr.Rmd` is failing because of some strange `sessionInfo()` error from `ComplexHeatmap`. Not sure what is causing this.

**Comment by raysinensis** [comment]
moving all except `clustifyr.Rmd` is probably my vote. Ryan made the overall tutorial look pretty nice, and does include basic visualizations.

not sure about the `sessionInfo()` error

**Comment by kriemo** [comment]
Ok sounds good. I can take care of moving the vignettes. Also, I added a webhook as requested by bioconductor, so once we bump the version the package will be rebuilt at the bioconductor contributions repo.

**Comment by kriemo** [comment]
Made some fixes, so now the package builds without errors There are some notes about code style that will likely need to be addressed and a warning about not depending on R 4.0. Can discuss more in #326
